bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/auth/client/oauth2/OAuthFactory.java

```diff
@@ -0,0 +1,67 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.core.auth.client.oauth2;
+
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.Nullable;
+
+/**
+ * The OAuth Factory interface
+ *
+ * @author Michael Bock - Initial Contribution
+ * @author Gary Tse - ESH Adaptation
+ */
+public interface OAuthFactory {
+
+    /**
+     * Creates a new oauth service. Use this method only once to obtain a handle and store
+     * this handle for further in a persistent storage container.
+     *
+     * @param tokenUrl the token url of the oauth provider. This is used for getting access token.
+     * @param authorizationUrl the authorization url of the oauth provider. This is used purely for generating
+     *            authorization code/ url.
+     * @param clientId the client id
+     * @param clientSecret the client secret (optional)
+     * @param scope the desired scope
+     * @param supportsBasicAuth whether the service supports basic authorization or the client id and client secret
+     *            should be
+     *            passed as form params
+     * @return a handle to the oauth service
+     */
+    String createOAuthClientService(@NonNull String tokenUrl, @Nullable String authorizationUrl,
+            @NonNull String clientId, @Nullable String clientSecret, @Nullable String scope,
+            @Nullable Boolean supportsBasicAuth);
+
+    /**
+     * Gets the oauth service for a given handle
+     *
+     * @param handle the handle to the oauth service
+     * @return the oauth service, or null of the handle does not exist
+     */
+    OAuthClientService getOAuthClientService(@NonNull String handle);
+
+    /**
+     * Deletes an oauth service, together with all persisted authorization data under this service.
+     * i.e. Removes all Access Tokens.
+     *
+     * @param handle the handle to the oauth service
+     */
+    void unregisterOAuthService(@NonNull String handle);
+
+    /**
+     * This deletes all oauth services, together with all persisted authorization/configuration data.
+     * !Use with extreme caution!
+     */
+    void removeAllServiceAndHandle();
```

Comment:
I am wondering if there is a better name for this, for example, `removeAllOAuthServices()` (since all the methods comprise the word "`OAuthService`", I think we can add it here as well.

Reply:
Fixed